// tamagui/dist/cjs/index.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
}, __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var src_exports = {};
__export(src_exports, {
  ComponentContext: () => import_core.ComponentContext,
  Configuration: () => import_core.Configuration,
  FontLanguage: () => import_core.FontLanguage,
  Spacer: () => import_core.Spacer,
  Stack: () => import_core.Stack,
  Theme: () => import_core.Theme,
  Unspaced: () => import_core.Unspaced,
  View: () => import_core.View,
  createComponent: () => import_core.createComponent,
  createFont: () => import_core.createFont,
  createShorthands: () => import_core.createShorthands,
  createStyledContext: () => import_core.createStyledContext,
  createTheme: () => import_core.createTheme,
  createTokens: () => import_core.createTokens,
  createVariable: () => import_core.createVariable,
  getConfig: () => import_core.getConfig,
  getMedia: () => import_core.getMedia,
  getStylesAtomic: () => import_core.getStylesAtomic,
  getThemes: () => import_core.getThemes,
  getToken: () => import_core.getToken,
  getTokenValue: () => import_core.getTokenValue,
  getTokens: () => import_core.getTokens,
  getVariable: () => import_core.getVariable,
  getVariableName: () => import_core.getVariableName,
  getVariableValue: () => import_core.getVariableValue,
  insertFont: () => import_core.insertFont,
  isChrome: () => import_core.isChrome,
  isClient: () => import_core.isClient,
  isServer: () => import_core.isServer,
  isTamaguiComponent: () => import_core.isTamaguiComponent,
  isTamaguiElement: () => import_core.isTamaguiElement,
  isTouchable: () => import_core.isTouchable,
  isVariable: () => import_core.isVariable,
  isWeb: () => import_core.isWeb,
  isWebTouchable: () => import_core.isWebTouchable,
  matchMedia: () => import_core.matchMedia,
  mediaObjectToString: () => import_core.mediaObjectToString,
  mediaQueryConfig: () => import_core.mediaQueryConfig,
  mediaState: () => import_core.mediaState,
  setupDev: () => import_core.setupDev,
  setupReactNative: () => import_core.setupReactNative,
  spacedChildren: () => import_core.spacedChildren,
  styled: () => import_core.styled,
  themeable: () => import_core.themeable,
  useConfiguration: () => import_core.useConfiguration,
  useDidFinishSSR: () => import_core.useDidFinishSSR,
  useEvent: () => import_core.useEvent,
  useGet: () => import_core.useGet,
  useIsTouchDevice: () => import_core.useIsTouchDevice,
  useIsomorphicLayoutEffect: () => import_core.useIsomorphicLayoutEffect,
  useMedia: () => import_core.useMedia,
  useProps: () => import_core.useProps,
  usePropsAndStyle: () => import_core.usePropsAndStyle,
  useStyle: () => import_core.useStyle,
  useTheme: () => import_core.useTheme,
  useThemeName: () => import_core.useThemeName,
  variableToString: () => import_core.variableToString,
  withStaticProperties: () => import_core.withStaticProperties
});
module.exports = __toCommonJS(src_exports);
var import_setup = require("./setup");
__reExport(src_exports, require("@tamagui/accordion"), module.exports);
__reExport(src_exports, require("@tamagui/adapt"), module.exports);
__reExport(src_exports, require("@tamagui/alert-dialog"), module.exports);
__reExport(src_exports, require("@tamagui/animate-presence"), module.exports);
__reExport(src_exports, require("@tamagui/avatar"), module.exports);
__reExport(src_exports, require("@tamagui/button"), module.exports);
__reExport(src_exports, require("@tamagui/card"), module.exports);
__reExport(src_exports, require("@tamagui/checkbox"), module.exports);
__reExport(src_exports, require("@tamagui/compose-refs"), module.exports);
__reExport(src_exports, require("@tamagui/create-context"), module.exports);
__reExport(src_exports, require("@tamagui/dialog"), module.exports);
__reExport(src_exports, require("@tamagui/font-size"), module.exports);
__reExport(src_exports, require("@tamagui/form"), module.exports);
__reExport(src_exports, require("@tamagui/group"), module.exports);
__reExport(src_exports, require("@tamagui/helpers-tamagui"), module.exports);
__reExport(src_exports, require("@tamagui/image"), module.exports);
__reExport(src_exports, require("@tamagui/label"), module.exports);
__reExport(src_exports, require("@tamagui/list-item"), module.exports);
__reExport(src_exports, require("@tamagui/popover"), module.exports);
__reExport(src_exports, require("@tamagui/popper"), module.exports);
__reExport(src_exports, require("@tamagui/portal"), module.exports);
__reExport(src_exports, require("@tamagui/progress"), module.exports);
__reExport(src_exports, require("@tamagui/radio-group"), module.exports);
__reExport(src_exports, require("@tamagui/scroll-view"), module.exports);
__reExport(src_exports, require("@tamagui/select"), module.exports);
__reExport(src_exports, require("@tamagui/separator"), module.exports);
__reExport(src_exports, require("@tamagui/shapes"), module.exports);
__reExport(src_exports, require("@tamagui/sheet"), module.exports);
__reExport(src_exports, require("@tamagui/slider"), module.exports);
__reExport(src_exports, require("@tamagui/stacks"), module.exports);
__reExport(src_exports, require("@tamagui/switch"), module.exports);
__reExport(src_exports, require("@tamagui/tabs"), module.exports);
__reExport(src_exports, require("@tamagui/text"), module.exports);
__reExport(src_exports, require("@tamagui/theme"), module.exports);
__reExport(src_exports, require("@tamagui/toggle-group"), module.exports);
__reExport(src_exports, require("@tamagui/tooltip"), module.exports);
__reExport(src_exports, require("@tamagui/use-controllable-state"), module.exports);
__reExport(src_exports, require("@tamagui/use-debounce"), module.exports);
__reExport(src_exports, require("@tamagui/use-force-update"), module.exports);
__reExport(src_exports, require("@tamagui/use-window-dimensions"), module.exports);
__reExport(src_exports, require("@tamagui/visually-hidden"), module.exports);
__reExport(src_exports, require("./createTamagui"), module.exports);
__reExport(src_exports, require("./viewTypes"), module.exports);
__reExport(src_exports, require("./views/TamaguiProvider"), module.exports);
__reExport(src_exports, require("./views/Anchor"), module.exports);
__reExport(src_exports, require("./views/EnsureFlexed"), module.exports);
__reExport(src_exports, require("./views/Fieldset"), module.exports);
__reExport(src_exports, require("./views/Input"), module.exports);
__reExport(src_exports, require("./views/Layouts"), module.exports);
__reExport(src_exports, require("./views/Spinner"), module.exports);
__reExport(src_exports, require("./views/TextArea"), module.exports);
__reExport(src_exports, require("./views/Text"), module.exports);
var import_core = require("@tamagui/core");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ComponentContext,
  Configuration,
  FontLanguage,
  Spacer,
  Stack,
  Theme,
  Unspaced,
  View,
  createComponent,
  createFont,
  createShorthands,
  createStyledContext,
  createTheme,
  createTokens,
  createVariable,
  getConfig,
  getMedia,
  getStylesAtomic,
  getThemes,
  getToken,
  getTokenValue,
  getTokens,
  getVariable,
  getVariableName,
  getVariableValue,
  insertFont,
  isChrome,
  isClient,
  isServer,
  isTamaguiComponent,
  isTamaguiElement,
  isTouchable,
  isVariable,
  isWeb,
  isWebTouchable,
  matchMedia,
  mediaObjectToString,
  mediaQueryConfig,
  mediaState,
  setupDev,
  setupReactNative,
  spacedChildren,
  styled,
  themeable,
  useConfiguration,
  useDidFinishSSR,
  useEvent,
  useGet,
  useIsTouchDevice,
  useIsomorphicLayoutEffect,
  useMedia,
  useProps,
  usePropsAndStyle,
  useStyle,
  useTheme,
  useThemeName,
  variableToString,
  withStaticProperties,
  ...require("@tamagui/accordion"),
  ...require("@tamagui/adapt"),
  ...require("@tamagui/alert-dialog"),
  ...require("@tamagui/animate-presence"),
  ...require("@tamagui/avatar"),
  ...require("@tamagui/button"),
  ...require("@tamagui/card"),
  ...require("@tamagui/checkbox"),
  ...require("@tamagui/compose-refs"),
  ...require("@tamagui/create-context"),
  ...require("@tamagui/dialog"),
  ...require("@tamagui/font-size"),
  ...require("@tamagui/form"),
  ...require("@tamagui/group"),
  ...require("@tamagui/helpers-tamagui"),
  ...require("@tamagui/image"),
  ...require("@tamagui/label"),
  ...require("@tamagui/list-item"),
  ...require("@tamagui/popover"),
  ...require("@tamagui/popper"),
  ...require("@tamagui/portal"),
  ...require("@tamagui/progress"),
  ...require("@tamagui/radio-group"),
  ...require("@tamagui/scroll-view"),
  ...require("@tamagui/select"),
  ...require("@tamagui/separator"),
  ...require("@tamagui/shapes"),
  ...require("@tamagui/sheet"),
  ...require("@tamagui/slider"),
  ...require("@tamagui/stacks"),
  ...require("@tamagui/switch"),
  ...require("@tamagui/tabs"),
  ...require("@tamagui/text"),
  ...require("@tamagui/theme"),
  ...require("@tamagui/toggle-group"),
  ...require("@tamagui/tooltip"),
  ...require("@tamagui/use-controllable-state"),
  ...require("@tamagui/use-debounce"),
  ...require("@tamagui/use-force-update"),
  ...require("@tamagui/use-window-dimensions"),
  ...require("@tamagui/visually-hidden"),
  ...require("./createTamagui"),
  ...require("./viewTypes"),
  ...require("./views/TamaguiProvider"),
  ...require("./views/Anchor"),
  ...require("./views/EnsureFlexed"),
  ...require("./views/Fieldset"),
  ...require("./views/Input"),
  ...require("./views/Layouts"),
  ...require("./views/Spinner"),
  ...require("./views/TextArea"),
  ...require("./views/Text")
});
