// @tamagui/text/dist/cjs/Paragraph.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var Paragraph_exports = {};
__export(Paragraph_exports, {
  Paragraph: () => Paragraph
});
module.exports = __toCommonJS(Paragraph_exports);
var import_web = require("@tamagui/web"), import_SizableText = require("./SizableText");
const Paragraph = (0, import_web.styled)(import_SizableText.SizableText, {
  name: "Paragraph",
  tag: "p",
  userSelect: "auto",
  color: "$color",
  size: "$true"
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Paragraph
});
