// @tamagui/collapsible/dist/jsx/Collapsible.native.js
import { AnimatePresence } from "@tamagui/animate-presence";
import { useControllableState } from "@tamagui/use-controllable-state";
import {
  Stack,
  composeEventHandlers,
  createStyledContext,
  styled,
  withStaticProperties
} from "@tamagui/web";
import * as React from "react";
const COLLAPSIBLE_NAME = "Collapsible", { Provider: CollapsibleProvider, useStyledContext: useCollapsibleContext } = createStyledContext(), _Collapsible = React.forwardRef(
  (props, forwardedRef) => {
    const {
      __scopeCollapsible,
      open: openProp,
      defaultOpen,
      disabled,
      onOpenChange,
      ...collapsibleProps
    } = props, [open = !1, setOpen] = useControllableState({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    return <CollapsibleProvider
      scope={__scopeCollapsible}
      disabled={disabled}
      contentId={React.useId()}
      open={open}
      onOpenToggle={React.useCallback(
        () => setOpen((prevOpen) => !prevOpen),
        [setOpen]
      )}
    ><Stack
      data-state={getState(open)}
      data-disabled={disabled ? "" : void 0}
      {...collapsibleProps}
      ref={forwardedRef}
    /></CollapsibleProvider>;
  }
);
_Collapsible.displayName = COLLAPSIBLE_NAME;
const TRIGGER_NAME = "CollapsibleTrigger", CollapsibleTriggerFrame = styled(Stack, {
  name: TRIGGER_NAME,
  tag: "button"
}), CollapsibleTrigger = CollapsibleTriggerFrame.styleable(
  (props, forwardedRef) => {
    const { __scopeCollapsible, children, ...triggerProps } = props, context = useCollapsibleContext(__scopeCollapsible);
    return <CollapsibleTriggerFrame
      aria-controls={context.contentId}
      aria-expanded={context.open || !1}
      data-state={getState(context.open)}
      data-disabled={context.disabled ? "" : void 0}
      disabled={context.disabled}
      {...triggerProps}
      ref={forwardedRef}
      onPress={composeEventHandlers(props.onPress, context.onOpenToggle)}
    >{typeof children == "function" ? children({ open: context.open }) : children}</CollapsibleTriggerFrame>;
  }
);
CollapsibleTrigger.displayName = TRIGGER_NAME;
const CONTENT_NAME = "CollapsibleContent", CollapsibleContentFrame = styled(Stack, {
  name: CONTENT_NAME
}), CollapsibleContent = CollapsibleContentFrame.styleable((props, forwardedRef) => {
  const { forceMount, children, __scopeCollapsible, ...contentProps } = props, context = useCollapsibleContext(__scopeCollapsible);
  return <AnimatePresence {...contentProps}>{forceMount || context.open ? <CollapsibleContentFrame ref={forwardedRef} {...contentProps}>{children}</CollapsibleContentFrame> : null}</AnimatePresence>;
});
CollapsibleContent.displayName = CONTENT_NAME;
function getState(open) {
  return open ? "open" : "closed";
}
const Collapsible = withStaticProperties(_Collapsible, {
  Trigger: CollapsibleTrigger,
  Content: CollapsibleContent
});
export {
  Collapsible,
  CollapsibleContent,
  CollapsibleContentFrame,
  CollapsibleTrigger,
  CollapsibleTriggerFrame
};
