// @tamagui/select/dist/jsx/SelectViewport.js
import { FloatingFocusManager } from "@floating-ui/react";
import { AnimatePresence } from "@tamagui/animate-presence";
import {
  composeRefs,
  isWeb,
  useIsomorphicLayoutEffect
} from "@tamagui/core";
import { styled } from "@tamagui/core";
import { PortalItem } from "@tamagui/portal";
import { ThemeableStack } from "@tamagui/stacks";
import { VIEWPORT_NAME } from "./constants";
import {
  ForwardSelectContext,
  useSelectContext,
  useSelectItemParentContext
} from "./context";
import { useSelectBreakpointActive } from "./useSelectBreakpointActive";
const SelectViewportFrame = styled(ThemeableStack, {
  name: VIEWPORT_NAME,
  variants: {
    unstyled: {
      false: {
        size: "$2",
        backgroundColor: "$background",
        elevate: !0,
        bordered: !0,
        userSelect: "none",
        outlineWidth: 0
      }
    },
    size: {
      "...size": (val, { tokens }) => ({
        borderRadius: tokens.radius[val] ?? val
      })
    }
  },
  defaultVariants: {
    unstyled: !1
  }
}), SelectViewport = SelectViewportFrame.styleable(
  function(props, forwardedRef) {
    const { __scopeSelect, children, disableScroll, ...viewportProps } = props, context = useSelectContext(VIEWPORT_NAME, __scopeSelect), itemContext = useSelectItemParentContext(VIEWPORT_NAME, __scopeSelect), breakpointActive = useSelectBreakpointActive(context.sheetBreakpoint);
    useIsomorphicLayoutEffect(() => {
      context.update && context.update();
    }, [breakpointActive]);
    const composedRefs = composeRefs(
      forwardedRef,
      context.floatingContext?.refs.setFloating
    );
    if (itemContext.shouldRenderWebNative)
      return <>{children}</>;
    if (breakpointActive || !isWeb)
      return <PortalItem hostName={`${context.scopeKey}SheetContents`}><ForwardSelectContext
        __scopeSelect={__scopeSelect}
        itemContext={itemContext}
        context={context}
      >{children}</ForwardSelectContext></PortalItem>;
    if (!itemContext.interactions)
      return process.env.NODE_ENV === "development" && console.warn("No interactions provided to Select, potentially missing Adapt"), null;
    const {
      style,
      // remove this, it was set to "Select" always
      className,
      ...floatingProps
    } = itemContext.interactions.getFloatingProps(), { scrollbarWidth, listStyleType, overflow, ...restStyle } = style;
    return <>
      {!disableScroll && !props.unstyled && <style
        dangerouslySetInnerHTML={{
          __html: selectViewportCSS
        }}
      />}
      <FloatingFocusManager context={context.floatingContext}><AnimatePresence>{context.open ? <SelectViewportFrame
        disableClassName
        key="select-viewport"
        size={itemContext.size}
        role="presentation"
        {...viewportProps}
        ref={composedRefs}
        {...floatingProps}
        {...restStyle}
        {...!props.unstyled && {
          overflow: disableScroll ? void 0 : overflow ?? "scroll"
        }}
      >{children}</SelectViewportFrame> : null}</AnimatePresence></FloatingFocusManager>
      {
        /* keep in dom to allow for portal to the trigger... very hacky! we should fix */
      }
      {!context.open && <div style={{ display: "none" }}>{props.children}</div>}
    </>;
  }
), selectViewportCSS = `
.is_SelectViewport {
  scrollbar-width: none;
  -webkit-overflow-scrolling: touch;
  overscroll-behavior: contain;
}

.is_SelectViewport::-webkit-scrollbar{
  display:none
}
`;
export {
  SelectViewport,
  SelectViewportFrame
};
