// @tamagui/web/dist/cjs/helpers/matchMedia.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var matchMedia_native_exports = {};
__export(matchMedia_native_exports, {
  matchMedia: () => matchMedia,
  setupMatchMedia: () => setupMatchMedia
});
module.exports = __toCommonJS(matchMedia_native_exports);
let matchMediaImpl = matchMediaFallback;
const matchMedia = (...args) => matchMediaImpl(...args);
function matchMediaFallback(query) {
  return process.env.NODE_ENV === "development" && console.warn("warning: matchMedia implementation is not provided."), {
    match: (a, b) => !1,
    addListener: () => {
    },
    removeListener: () => {
    },
    matches: !1
  };
}
function setupMatchMedia(_) {
  matchMediaImpl = _, globalThis.matchMedia = _;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  matchMedia,
  setupMatchMedia
});
