// @tamagui/web/dist/cjs/helpers/normalizeShadow.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var normalizeShadow_native_exports = {};
__export(normalizeShadow_native_exports, {
  normalizeShadow: () => normalizeShadow
});
module.exports = __toCommonJS(normalizeShadow_native_exports);
var import_normalize_css_color = require("@tamagui/normalize-css-color"), import_defaultOffset = require("./defaultOffset"), import_normalizeColor = require("./normalizeColor"), import_normalizeValueWithProperty = require("./normalizeValueWithProperty");
function normalizeShadow({
  shadowColor,
  shadowOffset,
  shadowOpacity,
  shadowRadius
}) {
  const { height, width } = shadowOffset || import_defaultOffset.defaultOffset, colorStr = String(shadowColor || "black"), val = (0, import_normalize_css_color.normalizeCSSColor)(colorStr);
  if (typeof val != "number")
    throw new Error(`invalid shadow color ${colorStr}`);
  const { r, g, b, a } = (0, import_normalizeColor.rgba)(val);
  return {
    shadowOffset: {
      width: (0, import_normalizeValueWithProperty.normalizeValueWithProperty)(width || 0),
      height: (0, import_normalizeValueWithProperty.normalizeValueWithProperty)(height || 0)
    },
    shadowRadius: (0, import_normalizeValueWithProperty.normalizeValueWithProperty)(shadowRadius || 0),
    shadowColor: `rgb(${r},${g},${b})`,
    shadowOpacity: shadowOpacity ?? a
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  normalizeShadow
});
