// @tamagui/switch/dist/cjs/SwitchContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var SwitchContext_exports = {};
__export(SwitchContext_exports, {
  SwitchContext: () => SwitchContext
});
module.exports = __toCommonJS(SwitchContext_exports);
var import_core = require("@tamagui/core");
const SwitchContext = (0, import_core.createStyledContext)({
  checked: !1,
  disabled: !1,
  size: void 0,
  frameWidth: 0,
  unstyled: !1
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SwitchContext
});
