// @tamagui/scroll-view/dist/cjs/ScrollView.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var ScrollView_exports = {};
__export(ScrollView_exports, {
  ScrollView: () => ScrollView
});
module.exports = __toCommonJS(ScrollView_exports);
var import_stacks = require("@tamagui/stacks"), import_web = require("@tamagui/web"), import_react_native = require("react-native");
const ScrollView = (0, import_web.styled)(import_react_native.ScrollView, {
  name: "ScrollView",
  scrollEnabled: !0,
  variants: {
    fullscreen: {
      true: import_stacks.fullscreenStyle
    }
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ScrollView
});
