// @tamagui/web/dist/esm/constants/accessibilityDirectMap.native.js
const accessibilityDirectMap = {}, webToNativeAccessibilityDirectMap = {
  "aria-label": "accessibilityLabel",
  "aria-labelledby": "accessibilityLabelledBy",
  "aria-live": "accessibilityLiveRegion",
  "aria-modal": "accessibilityViewIsModal",
  "aria-hidden": "accessibilityElementsHidden"
}, nativeAccessibilityValue = {
  "aria-valuemin": "min",
  "aria-valuemax": "max",
  "aria-valuenow": "now",
  "aria-valuetext": "text"
}, nativeAccessibilityState = {
  "aria-disabled": "disabled",
  "aria-selected": "selected",
  "aria-checked": "checked",
  "aria-busy": "busy",
  "aria-expanded": "expanded"
}, accessibilityWebRoleToNativeRole = {
  alert: "alert",
  button: "button",
  checkbox: "checkbox",
  combobox: "combobox",
  grid: "grid",
  group: "none",
  heading: "header",
  imagebutton: "imagebutton",
  img: "image",
  keyboardkey: "keyboardkey",
  link: "link",
  menu: "menu",
  menubar: "menubar",
  menuitem: "menuitem",
  none: "none",
  presentation: "none",
  progressbar: "progressbar",
  radio: "radio",
  radiogroup: "radiogroup",
  region: "summary",
  scrollbar: "scrollbar",
  searchbox: "search",
  slider: "adjustable",
  spinbutton: "spinbutton",
  summary: "summary",
  switch: "switch",
  tab: "tab",
  tablist: "tablist",
  text: "text",
  timer: "timer",
  togglebutton: "togglebutton",
  toolbar: "toolbar"
};
export {
  accessibilityDirectMap,
  accessibilityWebRoleToNativeRole,
  nativeAccessibilityState,
  nativeAccessibilityValue,
  webToNativeAccessibilityDirectMap
};
