// @tamagui/select/dist/cjs/constants.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var constants_exports = {};
__export(constants_exports, {
  FALLBACK_THRESHOLD: () => FALLBACK_THRESHOLD,
  MIN_HEIGHT: () => MIN_HEIGHT,
  SCROLL_ARROW_THRESHOLD: () => SCROLL_ARROW_THRESHOLD,
  SCROLL_ARROW_VELOCITY: () => SCROLL_ARROW_VELOCITY,
  SELECT_NAME: () => SELECT_NAME,
  VIEWPORT_NAME: () => VIEWPORT_NAME,
  WINDOW_PADDING: () => WINDOW_PADDING
});
module.exports = __toCommonJS(constants_exports);
const SELECT_NAME = "Select", WINDOW_PADDING = 8, SCROLL_ARROW_VELOCITY = 8, SCROLL_ARROW_THRESHOLD = 8, MIN_HEIGHT = 80, FALLBACK_THRESHOLD = 16, VIEWPORT_NAME = "SelectViewport";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FALLBACK_THRESHOLD,
  MIN_HEIGHT,
  SCROLL_ARROW_THRESHOLD,
  SCROLL_ARROW_VELOCITY,
  SELECT_NAME,
  VIEWPORT_NAME,
  WINDOW_PADDING
});
