// @tamagui/select/dist/cjs/context.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var context_exports = {};
__export(context_exports, {
  ForwardSelectContext: () => ForwardSelectContext,
  SelectItemParentProvider: () => SelectItemParentProvider,
  SelectProvider: () => SelectProvider,
  createSelectContext: () => createSelectContext,
  createSelectItemParentContext: () => createSelectItemParentContext,
  createSelectItemParentScope: () => createSelectItemParentScope,
  createSelectScope: () => createSelectScope,
  useSelectContext: () => useSelectContext,
  useSelectItemParentContext: () => useSelectItemParentContext
});
module.exports = __toCommonJS(context_exports);
var import_create_context = require("@tamagui/create-context"), import_constants = require("./constants"), import_jsx_runtime = require("react/jsx-runtime");
const [createSelectContext, createSelectScope] = (0, import_create_context.createContextScope)(import_constants.SELECT_NAME), [SelectProvider, useSelectContext] = createSelectContext(import_constants.SELECT_NAME), [createSelectItemParentContext, createSelectItemParentScope] = (0, import_create_context.createContextScope)(import_constants.SELECT_NAME), [SelectItemParentProvider, useSelectItemParentContext] = createSelectContext(import_constants.SELECT_NAME), ForwardSelectContext = ({
  __scopeSelect,
  context,
  itemContext,
  children
}) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SelectProvider, { isInSheet: !0, scope: __scopeSelect, ...context, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SelectItemParentProvider, { scope: __scopeSelect, ...itemContext, children }) });
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ForwardSelectContext,
  SelectItemParentProvider,
  SelectProvider,
  createSelectContext,
  createSelectItemParentContext,
  createSelectItemParentScope,
  createSelectScope,
  useSelectContext,
  useSelectItemParentContext
});
