// @tamagui/web/dist/cjs/constants/accessibilityDirectMap.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var accessibilityDirectMap_native_exports = {};
__export(accessibilityDirectMap_native_exports, {
  accessibilityDirectMap: () => accessibilityDirectMap,
  accessibilityWebRoleToNativeRole: () => accessibilityWebRoleToNativeRole,
  nativeAccessibilityState: () => nativeAccessibilityState,
  nativeAccessibilityValue: () => nativeAccessibilityValue,
  webToNativeAccessibilityDirectMap: () => webToNativeAccessibilityDirectMap
});
module.exports = __toCommonJS(accessibilityDirectMap_native_exports);
const accessibilityDirectMap = {}, webToNativeAccessibilityDirectMap = {
  "aria-label": "accessibilityLabel",
  "aria-labelledby": "accessibilityLabelledBy",
  "aria-live": "accessibilityLiveRegion",
  "aria-modal": "accessibilityViewIsModal",
  "aria-hidden": "accessibilityElementsHidden"
}, nativeAccessibilityValue = {
  "aria-valuemin": "min",
  "aria-valuemax": "max",
  "aria-valuenow": "now",
  "aria-valuetext": "text"
}, nativeAccessibilityState = {
  "aria-disabled": "disabled",
  "aria-selected": "selected",
  "aria-checked": "checked",
  "aria-busy": "busy",
  "aria-expanded": "expanded"
}, accessibilityWebRoleToNativeRole = {
  alert: "alert",
  button: "button",
  checkbox: "checkbox",
  combobox: "combobox",
  grid: "grid",
  group: "none",
  heading: "header",
  imagebutton: "imagebutton",
  img: "image",
  keyboardkey: "keyboardkey",
  link: "link",
  menu: "menu",
  menubar: "menubar",
  menuitem: "menuitem",
  none: "none",
  presentation: "none",
  progressbar: "progressbar",
  radio: "radio",
  radiogroup: "radiogroup",
  region: "summary",
  scrollbar: "scrollbar",
  searchbox: "search",
  slider: "adjustable",
  spinbutton: "spinbutton",
  summary: "summary",
  switch: "switch",
  tab: "tab",
  tablist: "tablist",
  text: "text",
  timer: "timer",
  togglebutton: "togglebutton",
  toolbar: "toolbar"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  accessibilityDirectMap,
  accessibilityWebRoleToNativeRole,
  nativeAccessibilityState,
  nativeAccessibilityValue,
  webToNativeAccessibilityDirectMap
});
