// @tamagui/tooltip/dist/cjs/Tooltip.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var Tooltip_native_exports = {};
__export(Tooltip_native_exports, {
  Tooltip: () => Tooltip
});
module.exports = __toCommonJS(Tooltip_native_exports);
var import_core = require("@tamagui/core");
const RenderChildren = (props) => props.children, RenderNull = (props) => null, Tooltip = (0, import_core.withStaticProperties)(RenderChildren, {
  Anchor: RenderChildren,
  Arrow: RenderNull,
  Close: RenderNull,
  Content: RenderNull,
  Trigger: RenderChildren
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Tooltip
});
