// @tamagui/toggle-group/dist/esm/Toggle.native.js
import { ThemeableStack } from "@tamagui/stacks";
import { useControllableState } from "@tamagui/use-controllable-state";
import { composeEventHandlers, styled } from "@tamagui/web";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
const NAME = "Toggle", ToggleFrame = styled(ThemeableStack, {
  name: NAME,
  tag: "button",
  variants: {
    unstyled: {
      false: {
        pressTheme: !0,
        backgroundColor: "$background",
        alignItems: "center",
        justifyContent: "center",
        display: "flex",
        borderColor: "$borderColor",
        borderWidth: 1,
        margin: -1,
        hoverStyle: {
          backgroundColor: "$backgroundHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress"
        },
        focusStyle: {
          borderColor: "$borderColorFocus",
          outlineColor: "$borderColorFocus",
          outlineWidth: 2,
          outlineStyle: "solid"
        }
      }
    },
    active: {
      true: {
        zIndex: 1,
        hoverStyle: {
          backgroundColor: "$background"
        },
        focusStyle: {
          borderColor: "$borderColorPress",
          backgroundColor: "$backgroundPress"
        }
      }
    },
    orientation: {
      horizontal: {
        flexDirection: "row",
        spaceDirection: "horizontal"
      },
      vertical: {
        flexDirection: "column",
        spaceDirection: "vertical"
      }
    }
  },
  defaultVariants: {
    unstyled: !1
  }
}), Toggle = ToggleFrame.extractable(
  React.forwardRef((props, forwardedRef) => {
    const {
      pressed: pressedProp,
      defaultPressed = !1,
      onPressedChange,
      ...buttonProps
    } = props, [pressed = !1, setPressed] = useControllableState({
      prop: pressedProp,
      onChange: onPressedChange,
      defaultProp: defaultPressed
    });
    return /* @__PURE__ */ jsx(
      ToggleFrame,
      {
        theme: pressed ? "active" : null,
        themeShallow: !0,
        active: props.unstyled ? void 0 : pressed,
        "aria-pressed": pressed,
        "data-state": pressed ? "on" : "off",
        "data-disabled": props.disabled ? "" : void 0,
        ...buttonProps,
        ref: forwardedRef,
        onPress: composeEventHandlers(props.onPress ?? void 0, (event) => {
          props.disabled || setPressed(!pressed);
        })
      }
    );
  })
);
Toggle.displayName = NAME;
export {
  Toggle,
  ToggleFrame
};
