// tamagui/dist/cjs/views/EnsureFlexed.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var EnsureFlexed_exports = {};
__export(EnsureFlexed_exports, {
  EnsureFlexed: () => EnsureFlexed
});
module.exports = __toCommonJS(EnsureFlexed_exports);
var import_core = require("@tamagui/core");
const EnsureFlexed = (0, import_core.styled)(import_core.Text, {
  opacity: 0,
  lineHeight: 0,
  height: 0,
  display: "flex",
  fontSize: 200,
  children: "wwwwwwwwwwwwwwwwwww",
  hoverStyle: {
    backgroundColor: "red"
  },
  pointerEvents: "none"
});
EnsureFlexed.isVisuallyHidden = !0;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EnsureFlexed
});
