// tamagui/dist/cjs/views/Layouts.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var Layouts_exports = {};
__export(Layouts_exports, {
  Article: () => Article,
  Aside: () => Aside,
  Footer: () => Footer,
  Header: () => Header,
  Main: () => Main,
  Nav: () => Nav,
  Section: () => Section
});
module.exports = __toCommonJS(Layouts_exports);
var import_core = require("@tamagui/core");
const Section = (0, import_core.styled)(import_core.Stack, {
  name: "Section",
  tag: "section",
  flexDirection: "column",
  accessibilityRole: "summary"
}), Article = (0, import_core.styled)(import_core.Stack, {
  name: "Article",
  tag: "article",
  flexDirection: "column"
}), Main = (0, import_core.styled)(import_core.Stack, {
  name: "Main",
  tag: "main",
  flexDirection: "column"
}), Header = (0, import_core.styled)(import_core.Stack, {
  name: "Header",
  tag: "header",
  accessibilityRole: "header",
  flexDirection: "column"
}), Aside = (0, import_core.styled)(import_core.Stack, {
  name: "Aside",
  tag: "aside",
  flexDirection: "column"
  // accessibilityRole: 'complementary',
}), Footer = (0, import_core.styled)(import_core.Stack, {
  name: "Footer",
  tag: "footer",
  flexDirection: "column"
  // accessibilityRole: 'contentinfo',
}), Nav = (0, import_core.styled)(import_core.Stack, {
  name: "Nav",
  tag: "nav",
  flexDirection: "column"
  // accessibilityRole: 'navigation',
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Article,
  Aside,
  Footer,
  Header,
  Main,
  Nav,
  Section
});
