// @tamagui/stacks/dist/cjs/ThemeableStack.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var ThemeableStack_exports = {};
__export(ThemeableStack_exports, {
  ThemeableStack: () => ThemeableStack,
  themeableVariants: () => themeableVariants
});
module.exports = __toCommonJS(ThemeableStack_exports);
var import_core = require("@tamagui/core"), import_Stacks = require("./Stacks"), import_variants = require("./variants");
const chromelessStyle = {
  backgroundColor: "transparent",
  borderColor: "transparent",
  shadowColor: "transparent",
  hoverStyle: {
    borderColor: "transparent"
  }
}, themeableVariants = {
  backgrounded: {
    true: {
      backgroundColor: "$background"
    }
  },
  radiused: import_variants.radiused,
  hoverTheme: import_variants.hoverTheme,
  pressTheme: import_variants.pressTheme,
  focusTheme: import_variants.focusTheme,
  circular: import_variants.circular,
  padded: import_variants.padded,
  elevate: import_variants.elevate,
  bordered: import_variants.bordered,
  transparent: {
    true: {
      backgroundColor: "transparent"
    }
  },
  chromeless: {
    true: chromelessStyle,
    all: {
      ...chromelessStyle,
      hoverStyle: chromelessStyle,
      pressStyle: chromelessStyle,
      focusStyle: chromelessStyle
    }
  }
}, ThemeableStack = (0, import_core.styled)(import_Stacks.YStack, {
  variants: themeableVariants
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ThemeableStack,
  themeableVariants
});
