// @tamagui/linear-gradient/dist/cjs/LinearGradient.shared.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var LinearGradient_shared_exports = {};
__export(LinearGradient_shared_exports, {
  LinearGradient: () => LinearGradient
});
module.exports = __toCommonJS(LinearGradient_shared_exports);
var import_core = require("@tamagui/core"), import_stacks = require("@tamagui/stacks"), import_linear_gradient = require("./linear-gradient"), import_jsx_runtime = require("react/jsx-runtime");
const LinearGradient = import_stacks.YStack.styleable(
  (propsIn, ref) => {
    var _a;
    const props = (0, import_core.useProps)(propsIn), { start, end, colors: colorsProp, locations, children, ...stackProps } = props, theme = (0, import_core.useTheme)(), colors = ((_a = props.colors) == null ? void 0 : _a.map((c) => {
      var _a2;
      return ((_a2 = theme[c]) == null ? void 0 : _a2.get("web")) ?? c;
    })) || [];
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LinearGradientFrame, { ref, ...stackProps, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_linear_gradient.LinearGradient,
      {
        start,
        end,
        colors,
        locations,
        style: absoluteFill,
        children
      }
    ) });
  }
), absoluteFill = {
  position: "absolute",
  top: 0,
  left: 0,
  right: 0,
  bottom: 0
}, LinearGradientFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: "LinearGradient",
  overflow: "hidden",
  position: "relative"
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LinearGradient
});
